crates/sui-core/src/transaction_input_checker.rs

```diff
@@ -120,18 +121,22 @@ where
     // TODO: We should be able to allow the same shared object to show up
     // in more than one SingleTransactionKind. We need to ensure that their
     // version number only increases once at the end of the Batch execution.
-    let mut owned_object_authenticators: HashSet<SuiAddress> = HashSet::new();
+    let mut object_owner_map = BTreeMap::new();
     for object in objects.iter().flatten() {
         if !object.is_immutable() {
             fp_ensure!(
-                owned_object_authenticators.insert(object.id().into()),
+                object_owner_map.insert(object.id(), object.owner).is_none(),
                 SuiError::InvalidBatchTransaction {
                     error: format!("Mutable object {} cannot appear in more than one single transactions in a batch", object.id()),
                 }
             );
         }
     }
 
+    // build a map of each object to its ancestor object (if it has one)
+    // returns an error on cycles, so this should not occur as storage should be free of cycles
+    let root_ancestor_map =
```

Comment:
This function is at the front line, and it should certainly be possible that some object in the input ownership chain doesn't exist.